### PR TITLE
#167165056 Validates User Signup Input

### PR DESCRIPTION
**Pull request by OdunayoOkebunmi**
### What does this PR do?

Creates a  middleware to validate user signup input

#### Description of Task to be completed?
GIVEN user supplies invalid parameters for signup
WHEN user hits the signup button
THEN user should receive a descriptive error message telling the user why such details are not acceptable.

User should be able to sign up via the route: `api/v1/auth/signup`
 
#### How should this be manually tested?
After cloning this repo, `cd` into it and do the following:

```bash
# fetch this branch
$ git fetch origin ft-user-sign-out-167165056

# install dependencies
$ npm install

# run unit tests
$ npm test

# start the development server
$ npm run server
```

When the server is running, using Postman, make a `POST` request to the route stated above

#### What are the relevant pivotal tracker stories?
[#167165056](https://www.pivotaltracker.com/story/show/167165056)

#### Screenshots (if appropriate)
Screenshot 2019-07-25 at 20 24 26

### CODE REVIEW

File: __tests__/inputValidation.test.js [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+      .post(`${url}/auth/signup`)
+      .send(mockUsers[3])
+      .end((err, res) => {
+        expect(res.status).eql(400);
+        expect(res.body).to.have.property('error');
+        expect(res.body.error[0]).eql('Password must be at least 8 characters');
+        done();
+      });
+  });
+  it('should successful signup a user if the parameters are valid', (done) => {
```
Comment (mekzy-o) [inline __tests__/inputValidation.test.js]: it(should **successfully** signup....). I think its an oversight.

File: middlewares/inputValidation.js [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+      .isEmpty({ ignore_whitespace: true })
+      .withMessage('Password is required')
+      .matches(/^(?=.*\d)(?=.*[a-z])(?=.*[A-Z])[0-9a-zA-Z]/, 'i')
+      .withMessage('Password must contain at least one uppercase letter, one lowercase letter and one numeric digit')
+      .trim()
+      .isLength({ min: 8 })
+      .withMessage('Password must be at least 8 characters'),
+    (req, res, next) => {
+      const errors = validationResult(req);
+      const errorMessage = [];
```
Comment (mekzy-o) [inline middlewares/inputValidation.js]: if you are displaying **one error message** at a time, then it's best you use an object instead of an array.

Comment (OdunayoOkebunmi) [reply middlewares/inputValidation.js]: Noted. Thanks

File: __tests__/inputValidation.test.js [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+    chai.request(app)
+      .post(`${url}/auth/signup`)
+      .send(mockUsers[4])
+      .end((err, res) => {
+        expect(res.status).eql(201);
+        expect(res.body).to.have.property('user');
+        done();
```
Comment (mazma1) [inline __tests__/inputValidation.test.js:71]: Make this test case more rounded like the others. Can you assert that the details in the user response object matches the mock request payload?

File: index.js [hunk 4]
```diff
@@ -19,38 +21,31 @@ app.use(cors());
...
-);
+app.use('/api/v1', routes);
+app.use('/api/v1/docs', swaggerUi.serve, swaggerUi.setup(docs));
 
 if (!isProduction) {
   app.use(errorhandler());
 }
 
-// catch 404 and forward to error handler
+// / catch 404 and forward to error handler
```
Comment (mazma1) [inline index.js]: Remove the trailing forward slash here

File: index.js [hunk 5]
```diff
@@ -19,38 +21,31 @@ app.use(cors());
...
 
-// error handlers
+// / error handlers
 
 // development error handler
 // will print stacktrace
 if (!isProduction) {
   app.use((err, req, res, next) => {
-    console.log(err.stack);
+    // console.log(err.stack);
```
Comment (mazma1) [inline index.js]: Why not just delete this if it is no more needed?

File: package.json [hunk 6]
```diff
@@ -5,24 +5,30 @@
   "main": "index.js",
   "scripts": {
     "server": "nodemon --exec babel-node .",
-    "test": "nyc --reporter=html --reporter=text mocha --exit --timeout 10000  --require @babel/register ./__tests__/*.js",
+    "pretest": " NODE_ENV=test npm run migrate:undo && cross-env NODE_ENV=test npm run migrate && cross-env NODE_ENV=test npm run seedall",
+    "test": "cross-env NODE_ENV=test nyc --reporter=html --reporter=text mocha --exit --timeout 10000  --require @babel/register ./__tests__/*.js",
     "coveralls": "nyc report --reporter=text-lcov | coveralls",
-    "migrate": "sequelize db:migrate",
+    "migrate": "cross-env NODE_ENV=test && sequelize db:migrate",
```
Comment (mazma1) [inline package.json]: Why do we need a test environment for migration here?

File: index.js (hunk 5 above)
Comment (OdunayoOkebunmi) [reply index.js]: Okay, noted

File: __tests__/inputValidation.test.js [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+  });
+  it('should successfully signup a user if the parameters are valid', (done) => {
+    chai.request(app)
+      .post(`${url}/auth/signup`)
+      .send(mockUsers[4])
+      .end((err, res) => {
+        expect(res.status).eql(201);
+        expect(res.body).to.have.property('user');
+        expect(res.body.user).to.have.property('firstName');
+        expect(res.body.user).to.have.property('lastName');
```
Comment (mazma1) [inline __tests__/inputValidation.test.js:67]: This is still not fixed. 

Like I said, it is not enough to check that a response has a property, you need to be sure that the value of that property is what you expect it to be. So check that the `mockUsers[4]`'s firlstname, lastname, etc match what was returned in the response.

Something like:
`expect(res.body.user.firstName).eql(mockUsers[4].firstName);`

Does this make sense now?

Comment (OdunayoOkebunmi) [reply __tests__/inputValidation.test.js:67]: Yes. It does, I’ll fix that now

**Comment by mazma1** [comment]
@OdunayoOkebunmi great work cleaning up this task!

**Pull request merged**